### DISCUSSION TITLE
Add c like switch (fallthrough tag)?

### DISCUSSION THREAD

**Discussion by Andersama** [odin-lang/Odin#1936] [Proposals]
New to the language. I started reading the documentation and then looking at the source code for the compiler. It seems like it'd be relatively straightforward to have a c-like switch if there isn't one already. Seeing how `#partial` looks like it ignores unhandled cases, I'm thinking a `#fallthrough` tag or `switch fallthrough`. In the event someone wants fallthrough behavior by default. A tag appears to be far easier to do.

To be fair most of the code I've put in switch blocks doesn't use fallthroughs, but every once in a while it just makes sense. While I'm incredibly used to ()'s being everywhere I do like the different variations of for and switch statements. I'm not exactly invested in the language, but since it seems like it's designed with as few keystrokes in mind, I think someone might like this.

```odin
// if for example you wanted to write a particular device
#fallthrough switch x%8 {
case 0: //do no thing
case 1: do_thing()
case 2: do_thing()
case 3: do_thing()
// and so on...
}
```

Relevant area in the parser for anyone who'd want to add it as a tag:
https://github.com/odin-lang/Odin/blob/7b539e3025907ab5b40762f90ca4f613cee7eab5/src/parser.cpp#L4689

I might build the compiler myself and see if I can't work out how to do this for fun.

**Comment by Kelimion** [comment]
You can write that as either:
```odin
switch n {
case 0: //do no thing
case 1, 2, 3: do_thing()
}
```
Or:
```odin
switch n {
case 0: //do no thing
case 1: fallthrough
case 2: fallthrough
case 3: do_thing()
}
```

**Reply by Kelimion** [reply]
As to your statement that the language is optimized for terseness, it's not. It's optimized for internal consistency and readability, aiding long-term maintainability. Optimizing for terseness gives you horrors like Perl.

**Reply by Andersama** [reply]
Huh, I was under the impression the switch block was more or less like an if else chain, since there were examples with binary comparisons. I'd have assumed the first example would mean:
```odin
switch n {
case 0:
case 3:
}
```

You mention that the language prefers consistency over terseness, but I find it funny you provided a far more terse representation of the example I provided, which also wasn't particularly long. I guess my example was bad because there's syntax for handling multiple cases in one line. I'm not immediately coming up with any ideas for an instance where there's a bunch of cases but each statement is different in some way. If there's a way to write something like that without a ton of fallthroughs it sounds like odin's got this covered.

**Reply by Kelimion** [reply]
Then only 0 and 3 would ever be hit, with any other value ignored.

**Reply by Andersama** [reply]
Hold on, shouldn't your second example have been
```odin
switch n {
case 0: //do no thing
case 1: do_thing()
    fallthrough
case 2: do_thing()
    fallthrough
case 3: do_thing()
}
```
?
Now I might just have to play around with odin to figure out what's going on. I don't understand the syntax.

**Reply by Kelimion** [reply]
> Hold on, shouldn't your second example have been
> 
> ```odin
> switch n {
> case 0: //do no thing
> case 1: do_thing()
>     fallthrough
> case 2: do_thing()
>     fallthrough
> case 3: do_thing()
> }
> ```
> 
> ? Now I might just have to play around with odin to figure out what's going on. I don't understand the syntax.

No, because then `do_thing` would be called 3 times if n == 1, twice if n == 2, and once if n == 3. In my 2nd example it calls `do_thing` once if `n` is one of 1, 2 or 3.

**Reply by Andersama** [reply]
Sorry forgot to get back to this, I might've screwed up the first example so it wasn't quite clear (I probably should've had case 0: do_thing()). That was the intended effect of the original example.

**Reply by Kelimion** [reply]
Duff's Device?
